### PR TITLE
[Feature]: Improve client constructors

### PR DESCRIPTION
**Pull request by stevensJourney**
# Overview

This POC updates the `PowerSyncDatabase` constructor to behave a bit more similarly to the Dart SDK (within some limitations). 

The constructor previously relied on an instance of `DBAdapter` to be supplied. This caused the creation of a `PowerSyncDatabase` to require being wrapped inside a factory which generated the `DBAdapter`. Typically this required the use of `RNQSPowerSyncDatabaseOpenFactory` in React Native and `WASQLitePowerSyncDatabaseOpenFactory` in Web to create the `DBAdapter` and return an instance of `PowerSyncDatabase`. 

It's now possible to construct `PowerSyncDatabase` using the options from the `AbstractPowerSyncOpenFactory` directly.

This is achieved by moving out the logic from `AbstractPowerSyncOpenFactory` (and it's platform variants) which created SQLite connections to a light-weight `SQLOpenFactory` which just opens a `DBAdapter`. The PowerSync open factories still use this same logic internally, but are now deprecated in favour of using the  `PowerSyncDatabase` constructor directly.

A PowerSyncDatabase client can now be constructed by either passing it a `DBAdapter` which will directly be used internally; a `SQLOpenFactory` which will be used to create a `DBAdapter`; or by passing paramaters which will be passed to a default `SQLOpenFactory`.

This results in users being able to create clients as follows:

```diff
- export const db = new WASQLitePowerSyncDatabaseOpenFactory({
-  dbFilename: 'example.db',
-  schema: AppSchema
- }).getInstance();
+ export const db = new PowerSyncDatabase({
+ schema: AppSchema,
+ database: {
+  dbFilename: 'example.db'
+ }
+ });
```

## Testing

All demos have been updated to use the new constructor method. The demos were started locally to verify everything works correctly. 

Development packages for the SDKs have been published and installed inside an isolated copy of the React Todolist application. The latest development`@powersync/web` package SDK client was shown to be compatible with the latest stable `@powersync/react` package. 

All unit tests were updated to use the constructor method for creating PowerSync clients. An additional test has been added to ensure that the deprecated factory methods still function correctly.

### CODE REVIEW

**Comment by kobiebotha** [comment]
🚀

**Review by Chriztiaan — APPROVED** [review]

File: .changeset/fuzzy-dots-shave.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+'@powersync/react-native': minor
+'@powersync/common': minor
+'@powersync/web': minor
+---
+
+Improved constructor behaviour of PowerSyncDatabase and logic for open factories
```
Comment (benitav) [inline .changeset/fuzzy-dots-shave.md:7]:
```suggestion
Improved constructor behavior of `PowerSyncDatabase` and logic for open factories. Deprecates `RNQSPowerSyncDatabaseOpenFactory` and `WASQLitePowerSyncDatabaseOpenFactory`.
```

**Review by DominicGBauer — APPROVED** [review]

**Review by Chriztiaan — APPROVED** [review]

**Comment by stevensJourney** [comment]
Update: Added overloading of the `PowerSyncDatabase` constructors for better TypeScript type checking.

Previously the union of `DBAdapter | SQLOpenFactory | SQLiteOpenOptions` would allow any of the three interface's fields in the `database` param. 
image


Now with overloads the suggestions are better
image

**Review by Chriztiaan — APPROVED** [review]

**Review by DominicGBauer — APPROVED** [review]
This makes me so happy!

**Pull request merged**